iroh-bytes/src/downloader.rs

```diff
@@ -454,540 +503,495 @@ struct Service<G: Getter, D: Dialer> {
+                    let _ = self.queue.pop_front();
+                    let info = self.requests.remove(&kind).expect("queued downloads exist");
+                    self.finalize_download(kind, info.intents, Err(DownloadError::NoProviders));
                 }
             }
+        }
+    }
+
+    fn next_step(&self, kind: &DownloadKind) -> NextStep {
+        if self
+            .concurrency_limits
+            .at_requests_capacity(self.active_requests.len())
+        {
+            return NextStep::Wait;
         };
 
-        // we tried to get a node to perform this request but didn't get one, so now this attempt
-        // is failed
-        if remaining_retries > 0 {
-            remaining_retries -= 1;
-            self.schedule_request(kind, remaining_retries, next_node, intents);
-        } else {
-            // check if this hash is needed in some form, otherwise remove it from providers
-            let hash = *kind.hash();
-            if !self.is_needed(hash) {
-                self.providers.remove(hash)
+        let mut candidates = self.providers.get_candidates(&kind.hash()).peekable();
+        if candidates.peek().is_none() {
+            return NextStep::OutOfProviders;
+        }
+
+        let mut available = vec![];
+        let mut currently_dialing = 0;
+        let mut has_exhausted = false;
+        let mut next_to_dial = None;
+        for node in candidates {
+            match self.node_state(node) {
+                NodeState::Connected(info) => {
+                    let active_requests = info.active_requests();
+                    if self
+                        .concurrency_limits
+                        .node_at_request_capacity(active_requests)
+                    {
+                        has_exhausted = true;
+                    } else {
+                        available.push((node, active_requests));
+                    }
+                }
+                NodeState::Dialing => {
+                    currently_dialing += 1;
+                }
+                NodeState::Disconnected => {
+                    if next_to_dial.is_none() {
+                        next_to_dial = Some(node);
+                    }
+                }
             }
-            // request can't be retried
-            for sender in intents.into_values() {
-                let _ = sender.send(Err(anyhow::anyhow!("download ran out of attempts")));
+        }
+
+        let has_dialing = currently_dialing > 0;
+
+        if !available.is_empty() {
+            available.sort_unstable_by_key(|(_node, req_count)| *req_count);
+            let (node, _) = available.last().expect("just checked");
+            NextStep::StartTransfer(**node)
+        } else if let Some(node) = next_to_dial {
+            let at_dial_capacity = has_dialing
+                && self
+                    .concurrency_limits
+                    .at_dials_per_hash_capacity(currently_dialing);
```

Comment:
Not sure I understand your comment correctly: My understanding is:
* We iterate over the providers for this download, and count the number of nodes from the provider list which are currently being dialed
* If this number is higher than `at_dials_per_hash_capacity`, we don't queue another dial
This seems correct to me?

Reply:
The confusing part to me is dialing _for this hash_, which we don't track (because I don't think we care in general). However this is tricky to verify and maybe we should try it as a best effort thing, more than a hard limit treated as an invariant., and document it as such.

Consider the following scenario:

- `at_dials_per_hash_capacity = 2`
- `ProviderMap := {HashX <-> {PeerA, PeerB}}`
- `PeerA` is being dialed (for `HashX`)
- `PeerB` is being dialed (for `HashX`)
- `HashX` is at capacity
- queue now `HashY`, with `PeerC`, which is useful for `HashX` as well
- `ProviderMap := {HashX <-> {PeerA, PeerB, PeerC}, HashY <-> {PeerC}}`
- dial `PeerC` for `HashY`, well within capacity
- `HashX` is now above capacity

Reply:
Right, I understand the problem now, thank you!

I am thinking about two potentially viable options:

* What would happen if we remove the `at_dials_per_hash_capacity` limit? My thought for adding it was to not use all available connections capacity for dialing a single hash. However, in the current code flow, we only ever dial providers for the current head of the queue. When one of the dials succeeds, we start a transfer with that node, and move to the next item in the queue. The remaining running dials will not be aborted, but once they succeed, their state will be idle, so if the next item needs to start dials, it will return `NextStep::DialQueuedDisconnect`, which will terminate idle nodes as needed. The remaining issue *can* be that if hash A starts many dials, they won't be aborted until each connection is established, so the next hash in the queue might have to wait until those  move to connected-and-idle state to abort them.
So what we could do is mark dials as `required`, and abort non-required dials before terminating idle connections on `NextStep::DialQueuedDisconnect`. This is fine, but we still might not want to start 50 parallel dials at all if we happen to have many providers for a hash. So, option B:

* We leave things exactly as they are! `at_dials_per_hash_capacity` is not checked as an invariant, but only at the codepoint you linked, so only to determine whether a new dial should be started or not. Which is fine - the dial will either succeed at the connection be "up for picks" for the next element in the queue. Or it will fail, which will free a slot in `at_dials_per_hash_capacity` for all hashes that care about the node.

WDYT?

Reply:
yes second option aligns with my suggestion: treat it as a best effort thing and document it

Reply:
I pushed a commit that adds more docs to `at_dials_per_hash_capacity` and the logic in `next_step`.